python/lsst/ts/m2com/mock/mock_control_closed_loop.py

```diff
@@ -475,6 +475,286 @@ def select_axial_hardpoints(location_axial_actuator, specific_axial_hardpoint):
+            Delta x position in meter.
+        y : `float`
+            Delta y position in meter.
+        z : `float`
+            Delta z position in meter.
+        rx : `float`
+            Delta x rotator in radian.
+        ry : `float`
+            Delta y rotator in radian.
+        rz : `float`
+            Delta z rotator in radian.
+
+        Returns
+        -------
+        `numpy.ndarray`
+            All actuator displacements in meter.
+        """
+
+        # Consider the displacement from (x, y, z)
+
+        # Transformation matrix. The raw is each actuator's displacement. The
+        # column is the (x, y, z) movement.
+        matrix_trans = np.zeros((NUM_ACTUATOR, 3))
+
+        # A positive axial displacement results in a negative change in piston.
+        num_axial_actuators = NUM_ACTUATOR - NUM_TANGENT_LINK
+        matrix_trans[:num_axial_actuators, 2] = -1
+
+        for idx, angle in enumerate(location_tangent_link):
+            matrix_trans[num_axial_actuators + idx, 0] = np.cos(np.deg2rad(angle))
+            matrix_trans[num_axial_actuators + idx, 1] = -np.sin(np.deg2rad(angle))
+
+        # Translate the (x, y, z) motion to 78 actuator displacement.
+        disp_xyz = matrix_trans.dot(np.array([x, y, z]))
+
+        # Consider the displacement from (rx, ry, rz)
+
+        location_axial = np.append(
+            np.array(location_axial_actuator),
+            np.zeros((num_axial_actuators, 1)),
+            axis=1,
+        )
+
+        # Rotation matrix for the axial actuators
+        rot_x = np.array(
+            [[1, 0, 0], [0, np.cos(rx), np.sin(rx)], [0, -np.sin(rx), np.cos(rx)]]
+        )
+        rot_y = np.array(
+            [[np.cos(ry), 0, -np.sin(ry)], [0, 1, 0], [np.sin(ry), 0, np.cos(ry)]]
+        )
+        rot_z = np.array(
+            [[np.cos(rz), np.sin(rz), 0], [-np.sin(rz), np.cos(rz), 0], [0, 0, 1]]
+        )
+        rot_xyz = rot_x.dot(rot_y).dot(rot_z)
+
+        disp_rxryrz_axial = rot_xyz.dot(location_axial.T)
+        disp_rz_axial = disp_rxryrz_axial[2, :]
+
+        # Displacement rz, note there is a "-1" here for the direction of
+        # coordination system.
+        disp_rz_tangent = -radius * np.sin(rz) * np.ones(len(location_tangent_link))
+
+        disp_rz = np.append(disp_rz_axial, disp_rz_tangent)
+
+        return disp_xyz + disp_rz
+
+    @staticmethod
+    def hardpoint_to_rigid_body(
+        location_axial_actuator,
+        location_tangent_link,
+        radius,
+        hardpoints,
+        disp_hardpoint_current,
+        disp_hardpoint_home,
+    ):
+        """Calculate the rigid body position based on the hardpoint
+        displacements.
+
+        Notes
+        -----
+        Translate the calculation from the HardPointToRigidBody.vi in ts_mtm2.
+
+        Parameters
+        ----------
+        location_axial_actuator : `list [list]`
+            Location of the axial actuators: (x, y) in meter. This should be a
+            72 x 2 matrix.
```

Comment:
Same comment as above about the 72...

Reply:
Fixed in the commit: [779a5ae](https://github.com/lsst-ts/ts_m2com/pull/23/commits/779a5aed9b8f98cb5168e0eb012f54caa127766d). Thanks!